controller/rollout_save_controller.go

```diff
@@ -0,0 +1,1091 @@
+
+	}
+	clusterRoleList.Items = filtered
+
+	return getObjectsAndPrintToYAML(dir, name, func() (runtime.Object, error) {
+		return clusterRoleList, nil
+	}, scheme)
+}
+
+func (c *RolloutSaveController) generateRolloutYAMLsForClusterRoleBindingsByServiceAccounts(
+	clusterRoleBindingList *rbacv1.ClusterRoleBindingList,
+	serviceAccountList *corev1.ServiceAccountList,
+	dir, name string, scheme *runtime.Scheme) (err error) {
+	filtered := []rbacv1.ClusterRoleBinding{}
+	for _, clusterRoleBinding := range clusterRoleBindingList.Items {
+		doRollout := false
+		for _, serviceAccount := range serviceAccountList.Items {
+			for _, subject := range clusterRoleBinding.Subjects {
+				if subject.Kind != types.KubernetesKindServiceAccount {
+					continue
+				}
+				if subject.Name != serviceAccount.Name {
+					continue
+				}
+				if subject.Namespace != serviceAccount.Namespace {
+					continue
+				}
+				doRollout = true
+				break
+			}
+			if doRollout {
+				break
+			}
+		}
+		if doRollout {
+			filtered = append(filtered, clusterRoleBinding)
+		}
+	}
+	clusterRoleBindingList.Items = filtered
+
+	return getObjectsAndPrintToYAML(dir, name, func() (runtime.Object, error) {
+		return clusterRoleBindingList, nil
+	}, scheme)
+}
+
+func (c *RolloutSaveController) generateRolloutYAMLsForAPIExtensions(dir string) (err error) {
+	scheme := runtime.NewScheme()
+	err = apiextensionsv1.AddToScheme(scheme)
+	if err != nil {
+		return errors.Wrapf(err, "failed to add API Extension to scheme")
+	}
+
+	return getObjectsAndPrintToYAML(dir, "customresourcedefinitions", c.ds.GetAllLonghornCustomResourceDefinitions, scheme)
+}
+
+type GetRuntimeObjectListFunc func() (runtime.Object, error)
+
+func getObjectsAndPrintToYAML(dir, name string, getListFunc GetRuntimeObjectListFunc, scheme *runtime.Scheme) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "failed to generate %v", name)
+	}()
+
+	obj, err := getListFunc()
+	if err != nil {
+		return
+	}
+
+	err = addTypeInformationToObject(obj, scheme)
+	if err != nil {
+		return
+	}
+
+	path := filepath.Join(dir, name+".yaml")
+	err = os.MkdirAll(filepath.Dir(path), os.FileMode(0755))
+	if err != nil {
+		return
+	}
+
+	f, err := os.Create(path)
+	if err != nil {
+		return
+	}
+	defer f.Close()
+
+	printer := printers.YAMLPrinter{}
+	err = printer.PrintObj(obj, f)
+	if err != nil {
+		return
+	}
+
+	return nil
+}
+
+func addTypeInformationToObject(obj runtime.Object, scheme *runtime.Scheme) error {
+	gvks, _, err := scheme.ObjectKinds(obj)
+	if err != nil {
+		return errors.Wrapf(err, "failed to set ObjectKind, could missing apiVersion or kind and cannot assign it")
+	}
+
+	for _, gvk := range gvks {
+		if len(gvk.Kind) == 0 {
+			continue
+		}
+		if len(gvk.Version) == 0 || gvk.Version == runtime.APIVersionInternal {
+			continue
+		}
+		obj.GetObjectKind().SetGroupVersionKind(gvk)
+		break
+	}
+
+	return nil
+}
+
+func encodeToYAMLFile(obj interface{}, path string) (err error) {
```

Comment:
IIUC, we have the [Support bundle enhancement](https://github.com/longhorn/longhorn/issues/2759) in v1.4.0. Do we still need to enhance the support bundle functions?

Reply:
1.5 is the same as 1500m... As long as the value is recorded, I am fine with it.

Reply:
I mean I couldn't get the DaemonSet `requests.cpu` to equal exactly `1.5` before the RolloutSave generates the files. The `requests.cpu` seems to get automatically converted to `1500m` so that is what gets saved to the rollout file. And rollout restores `1500m` ok.